src/qc.rs

```diff
@@ -142,3 +142,123 @@ macro_rules! arbitrary_large {
 
 arbitrary_large!(I64: i64);
 arbitrary_large!(U64: u64);
+
+// Convenience macro to test intrinsics against their reference implementations.
+//
+// Each intrinsic is tested against both the `gcc_s` library as well as
+// `compiler-rt`. These libraries are defined in the `gcc_s` crate as well as
+// the `compiler-rt` crate in this repository. Both load a dynamic library and
+// lookup symbols through that dynamic library to ensure that we're using the
+// right intrinsic.
+//
+// This macro hopefully allows you to define a bare minimum of how to test an
+// intrinsic without worrying about these implementation details. A sample
+// invocation looks like:
+//
+//
+//    check! {
+//        // First argument is the function we're testing (either from this lib
+//        // or a dynamically loaded one. Further arguments are all generated by
+//        // quickcheck.
+//        fn __my_intrinsic(f: extern fn(i32) -> i32,
+//                          a: I32)
+//                          -> Option<(i32, i64)> {
+//
+//            // Discard tests by returning Some
+//            if a.0 == 0 {
+//                return None
+//            }
+//
+//            // Return the result via `Some` if the test can run
+//            let mut other_result = 0;
+//            let result = f(a.0, &mut other_result);
+//            Some((result, other_result))
+//        }
+//    }
+//
+// If anything returns `None` then the test is discarded, otherwise the two
+// results are compared for equality and the test fails if this equality check
+// fails.
+macro_rules! check {
```

Comment:
This is pretty much what I had in mind for #72. From the looks of things I'm glad I didn't have to write it 😆.

@alexcrichton do you have any thoughts about benchmarks? I'm assuming the starting performance target is "not significantly worse", which shouldn't be too bad since most of the code is pretty direct translation. I still think some basic benchmarks would be a good sanity check just to ward off possible regressions. I don't know how big the overhead is for dynamic function lookup but we could do that for `rustc-builtins` as well to make it more 'fair', I suppose.